acorn/src/expression.js

```diff
@@ -178,17 +182,23 @@ pp.parseExprOps = function(noIn, refDestructuringErrors) {
 // defer further parser to one of its callers when it encounters an
 // operator that has a lower precedence than the set it is parsing.
 
-pp.parseExprOp = function(left, leftStartPos, leftStartLoc, minPrec, noIn) {
+pp.parseExprOp = function(left, leftStartPos, leftStartLoc, minPrec, noIn, shortCircuitOps) {
   let prec = this.type.binop
   if (prec != null && (!noIn || this.type !== tt._in)) {
+    let logical = this.type === tt.logicalOR || this.type === tt.logicalAND
+    let coalesce = this.type === tt.coalesce
+    if (shortCircuitOps === ONLY_LOGICAL && coalesce || shortCircuitOps === ONLY_COALESCE && logical) {
+      this.raiseRecoverable(this.start, "Logical expressions and coalesce expressions cannot be mixed. Wrap either by parentheses")
```

Comment:
@mysticatea Why complicate this, and not only pass the token after `this.next()` as func arg? Then you could do a comparison on current token and the next token in the stream. You shouldn't need the constants. I did similar as suggested here in my locale code 
`if (((sC & Token.IsLogical) | (t & Token.IsCoalesc) | ((t & Token.IsLogical) | (sC & Token.IsCoalesc))) >    Token.IsCoalesc ) {`

Reply:
I'm not sure about your code. Would you elaborate?

Reply:
Let the shortCircuitOps constant be current token. After `this.next()` you pass **current** token as the `shortCircuitOps` arg. Then you get two tokens to compare against - token before and after you prime the scanner. 
``if(t1 === Coalece && t2 ===  Logical || t2 === Coalece t1 === Logical`. Then throw if match.
Actually you shouldn't need to pass anything here as function arg. You can store `t1` as a var and do the check after you prime the scanner before you call the `parseExprOp`

Reply:
The reason to pass constants is that I didn't want to do `t === tt.logicalOR || t === tt.logicalAND` repeatedly. We can do just `shortCircuitOps === ONLY_LOGICAL` by constants.

> Actually you shouldn't need to pass anything here as function arg.

How do you distinguish parentheses? For example, `a || b ?? c` is invalid, but `(a || b) ?? c` is valid.

Reply:
> How do you distinguish parentheses?
It happen by itself. If the token after `this.next()` is either logical or coalece you know that it isn't inside a paren. The comparison will not throw if either `t1` or `t2` is wrapped in a paren.

Reply:
@mysticatea Here is a [gist](https://gist.github.com/KFlash/7c708a597ca7a2f02894ddecd3e1cf30) from my private code that do the bin expr parsing. Maybe it can help. It pass Test262.

Reply:
Thank you for your advice. I simplified the logic. Does the new code express your intention?

Reply:
Looks good now, and should perform faster 👍